File: archive/pulumi-ts/Pr5379/lib/apigateway.ts
/**
 * API Gateway with WAF, throttling, and X-Ray
 */
import * as aws from '@pulumi/aws';
import * as pulumi from '@pulumi/pulumi';

export interface ApiGatewayArgs {
  apiName: string;
  lambdaFunction: aws.lambda.Function;
  wafWebAclArn: pulumi.Input<string>;
  environmentSuffix: string;
  tags: pulumi.Input<{ [key: string]: string }>;
}

export class ApiGatewayMigration extends pulumi.ComponentResource {
  public readonly api: aws.apigateway.RestApi;
  public readonly deployment: aws.apigateway.Deployment;
  public readonly stage: aws.apigateway.Stage;

  constructor(
    name: string,
    args: ApiGatewayArgs,
    opts?: pulumi.ComponentResourceOptions
  ) {
    super('tap:migration:ApiGatewayMigration', name, {}, opts);

    // Create REST API
    this.api = new aws.apigateway.RestApi(
      `${args.apiName}-${args.environmentSuffix}`,
      {
        name: `${args.apiName}-${args.environmentSuffix}`,
        description: 'Payment processing API',
        tags: args.tags,
      },
      { parent: this }
    );

    // Create resource
    const resource = new aws.apigateway.Resource(
      `${args.apiName}-resource-${args.environmentSuffix}`,
      {
        restApi: this.api.id,
        parentId: this.api.rootResourceId,
        pathPart: 'payment',
      },
      { parent: this }
    );

    // Create method
    const method = new aws.apigateway.Method(
      `${args.apiName}-method-${args.environmentSuffix}`,
      {
        restApi: this.api.id,
        resourceId: resource.id,
        httpMethod: 'POST',
        authorization: 'NONE',
      },
      { parent: this }
    );

    // Lambda integration
    const integration = new aws.apigateway.Integration(
      `${args.apiName}-integration-${args.environmentSuffix}`,
      {
        restApi: this.api.id,
        resourceId: resource.id,
        httpMethod: method.httpMethod,
        integrationHttpMethod: 'POST',
        type: 'AWS_PROXY',
        uri: args.lambdaFunction.invokeArn,
      },
      { parent: this }
    );

    // Lambda permission
    new aws.lambda.Permission(
      `${args.apiName}-lambda-permission-${args.environmentSuffix}`,
      {
        action: 'lambda:InvokeFunction',
        function: args.lambdaFunction.name,
        principal: 'apigateway.amazonaws.com',
        sourceArn: pulumi.interpolate`${this.api.executionArn}/*/*`,
      },
      { parent: this }
    );

    // Deployment
    this.deployment = new aws.apigateway.Deployment(
      `${args.apiName}-deployment-${args.environmentSuffix}`,
      {
        restApi: this.api.id,
        triggers: {
          redeployment: pulumi
            .all([resource.id, method.id, integration.id])
            .apply(ids => JSON.stringify(ids)),
        },
      },
      { parent: this, dependsOn: [method, integration] }
    );

    // Stage with X-Ray tracing
    this.stage = new aws.apigateway.Stage(
      `${args.apiName}-stage-${args.environmentSuffix}`,
      {
        restApi: this.api.id,
        deployment: this.deployment.id,
        stageName: 'prod',
        xrayTracingEnabled: true,
        tags: args.tags,
      },
      { parent: this }
    );

    // CloudWatch log group
    const logGroup = new aws.cloudwatch.LogGroup(
      `api-gateway-${args.apiName}-${args.environmentSuffix}`,
      {
        name: `/aws/apigateway/${args.apiName}-${args.environmentSuffix}`,
        retentionInDays: 30,
        tags: args.tags,
      },
      { parent: this }
    );

    // Method settings for throttling and logging
    new aws.apigateway.MethodSettings(
      `${args.apiName}-method-settings-${args.environmentSuffix}`,
      {
        restApi: this.api.id,
        stageName: this.stage.stageName,
        methodPath: '*/*',
        settings: {
          throttlingRateLimit: 1000,
          throttlingBurstLimit: 2000,
          loggingLevel: 'INFO',
          dataTraceEnabled: true,
          metricsEnabled: true,
        },
      },
      { parent: this, dependsOn: [this.stage, logGroup] }
    );

    // Associate WAF with API Gateway
    new aws.wafv2.WebAclAssociation(
      `${args.apiName}-waf-association-${args.environmentSuffix}`,
      {
        resourceArn: this.stage.arn,
        webAclArn: args.wafWebAclArn,
      },
      { parent: this, dependsOn: [this.stage] }
    );

    this.registerOutputs({
      apiId: this.api.id,
      apiEndpoint: this.stage.invokeUrl,
    });
  }
}

File: archive/pulumi-py/Pr5658/lib/infrastructure/api_gateway.py
"""
API Gateway module with usage plan and rate limiting.

This module creates API Gateway with proper Lambda integration,
usage plans, rate limiting, and X-Ray tracing.
"""

import pulumi
import pulumi_aws as aws
from pulumi import Output

from .aws_provider import AWSProviderManager
from .config import CICDPipelineConfig
from .lambda_functions import LambdaStack


class APIGatewayStack:
    """
    Manages API Gateway resources.
    
    Creates REST API with Lambda integration, usage plan,
    rate limiting, and X-Ray tracing.
    """
    
    def __init__(
        self,
        config: CICDPipelineConfig,
        provider_manager: AWSProviderManager,
        lambda_stack: LambdaStack
    ):
        """
        Initialize the API Gateway stack.
        
        Args:
            config: CICDPipelineConfig instance
            provider_manager: AWSProviderManager instance
            lambda_stack: LambdaStack instance
        """
        self.config = config
        self.provider_manager = provider_manager
        self.lambda_stack = lambda_stack
        self.api = None
        self.deployment = None
        self.stage = None
        
        self._create_api()
        self._create_resources()
        self._create_deployment()
        self._create_usage_plan()
    
    def _create_api(self):
        """Create REST API."""
        api_name = self.config.get_resource_name('api')
        
        self.api = aws.apigateway.RestApi(
            'api',
            name=api_name,
            description=f'API Gateway for {self.config.project_name}',
            endpoint_configuration=aws.apigateway.RestApiEndpointConfigurationArgs(
                types='REGIONAL'
            ),
            tags={
                **self.config.get_common_tags(),
                'Name': api_name
            },
            opts=self.provider_manager.get_resource_options()
        )
    
    def _create_resources(self):
        """Create API resources and methods."""
        function_name = 'pipeline-handler'
        function = self.lambda_stack.get_function(function_name)
        
        resource = aws.apigateway.Resource(
            'api-resource',
            rest_api=self.api.id,
            parent_id=self.api.root_resource_id,
            path_part='pipeline',
            opts=self.provider_manager.get_resource_options()
        )
        
        post_method = aws.apigateway.Method(
            'post-method',
            rest_api=self.api.id,
            resource_id=resource.id,
            http_method='POST',
            authorization='NONE',
            opts=self.provider_manager.get_resource_options()
        )
        
        post_integration = aws.apigateway.Integration(
            'post-integration',
            rest_api=self.api.id,
            resource_id=resource.id,
            http_method=post_method.http_method,
            integration_http_method='POST',
            type='AWS_PROXY',
            uri=function.invoke_arn,
            opts=self.provider_manager.get_resource_options()
        )
        
        get_method = aws.apigateway.Method(
            'get-method',
            rest_api=self.api.id,
            resource_id=resource.id,
            http_method='GET',
            authorization='NONE',
            opts=self.provider_manager.get_resource_options()
        )
        
        get_integration = aws.apigateway.Integration(
            'get-integration',
            rest_api=self.api.id,
            resource_id=resource.id,
            http_method=get_method.http_method,
            integration_http_method='POST',
            type='AWS_PROXY',
            uri=function.invoke_arn,
            opts=self.provider_manager.get_resource_options()
        )
        
        Output.all(self.api.execution_arn, resource.path).apply(
            lambda args: aws.lambda_.Permission(
                'api-lambda-permission',
                statement_id='AllowAPIGatewayInvoke',
                action='lambda:InvokeFunction',
                function=function.name,
                principal='apigateway.amazonaws.com',
                source_arn=f'{args[0]}/*/*/*',
                opts=self.provider_manager.get_resource_options()
            )
        )
        
        self.methods = [post_method, get_method]
        self.integrations = [post_integration, get_integration]
    
    def _create_deployment(self):
        """Create API deployment and stage."""
        deployment_name = self.config.get_resource_name('deployment')
        stage_name = 'v1'
        
        self.deployment = aws.apigateway.Deployment(
            'api-deployment',
            rest_api=self.api.id,
            triggers={
                'redeployment': Output.all(*[m.id for m in self.methods]).apply(
                    lambda ids: '-'.join(ids)
                )
            },
            opts=pulumi.ResourceOptions(
                provider=self.provider_manager.get_provider(),
                depends_on=self.methods + self.integrations
            )
        )
        
        log_group_name = f'/aws/apigateway/{self.config.get_resource_name("api")}'
        
        log_group = aws.cloudwatch.LogGroup(
            'api-log-group',
            name=log_group_name,
            retention_in_days=self.config.log_retention_days,
            tags={
                **self.config.get_common_tags(),
                'Name': log_group_name
            },
            opts=self.provider_manager.get_resource_options()
        )
        
        self.stage = aws.apigateway.Stage(
            'api-stage',
            rest_api=self.api.id,
            deployment=self.deployment.id,
            stage_name=stage_name,
            xray_tracing_enabled=self.config.enable_xray_tracing,
            tags={
                **self.config.get_common_tags(),
                'Name': self.config.get_resource_name(f'api-{stage_name}')
            },
            opts=self.provider_manager.get_resource_options()
        )
        
        aws.apigateway.MethodSettings(
            'api-method-settings',
            rest_api=self.api.id,
            stage_name=self.stage.stage_name,
            method_path='*/*',
            settings=aws.apigateway.MethodSettingsSettingsArgs(
                metrics_enabled=True,
                throttling_burst_limit=self.config.api_burst_limit,
                throttling_rate_limit=self.config.api_rate_limit
            ),
            opts=self.provider_manager.get_resource_options()
        )
    
    def _create_usage_plan(self):
        """Create usage plan with rate limiting."""
        usage_plan_name = self.config.get_resource_name('usage-plan')
        
        usage_plan = aws.apigateway.UsagePlan(
            'usage-plan',
            name=usage_plan_name,
            api_stages=[aws.apigateway.UsagePlanApiStageArgs(
                api_id=self.api.id,
                stage=self.stage.stage_name
            )],
            quota_settings=aws.apigateway.UsagePlanQuotaSettingsArgs(
                limit=self.config.api_quota_limit,
                period=self.config.api_quota_period
            ),
            throttle_settings=aws.apigateway.UsagePlanThrottleSettingsArgs(
                burst_limit=self.config.api_burst_limit,
                rate_limit=self.config.api_rate_limit
            ),
            tags={
                **self.config.get_common_tags(),
                'Name': usage_plan_name
            },
            opts=self.provider_manager.get_resource_options()
        )
        
        api_key_name = self.config.get_resource_name('api-key')
        
        api_key = aws.apigateway.ApiKey(
            'api-key',
            name=api_key_name,
            enabled=True,
            tags={
                **self.config.get_common_tags(),
                'Name': api_key_name
            },
            opts=self.provider_manager.get_resource_options()
        )
        
        aws.apigateway.UsagePlanKey(
            'usage-plan-key',
            key_id=api_key.id,
            key_type='API_KEY',
            usage_plan_id=usage_plan.id,
            opts=self.provider_manager.get_resource_options()
        )
        
        self.usage_plan = usage_plan
        self.api_key = api_key
    
    def get_api_id(self) -> Output[str]:
        """Get API ID."""
        return self.api.id
    
    def get_api_url(self) -> Output[str]:
        """Get API URL."""
        return Output.all(self.api.id, self.stage.stage_name).apply(
            lambda args: f'https://{args[0]}.execute-api.{self.config.primary_region}.amazonaws.com/{args[1]}'
        )
    
    def get_api_key_value(self) -> Output[str]:
        """Get API key value."""
        return self.api_key.value if self.api_key else Output.from_input('')

